array_analyzer/workflows/registration_workflow.py

```diff
@@ -96,20 +98,43 @@ def point_registration(input_dir, output_dir):
             stop_criteria=0.1,
             debug=constants.DEBUG,
         )
+
+        # Warn if fit is still bad
         if min_dist > constants.REG_DIST_THRESH:
-            warnings.warn("Registration failed, repeat with outlier removal")
+            spot_coords, im_filtered = img_processing.get_spot_coords(
+                image,
+                min_area=200,
+                min_thresh=100,
+                max_thresh=255,
+                min_circularity=0.1,
+                min_convexity=0.5,
+                min_dist_between_blobs=10,
+                min_repeatability=2,
+                sigma_gauss=10,
+                nbr_expected_spots=nbr_expected_spots,
+            )
             t_matrix, min_dist = registration.particle_filter(
                 fiducial_coords=fiducial_coords,
                 spot_coords=spot_coords,
                 particles=particles,
                 stds=np.array(constants.STDS),
                 stop_criteria=0.1,
-                remove_outlier=True,
                 debug=constants.DEBUG,
             )
-            # Warn if fit is still bad
             if min_dist > constants.REG_DIST_THRESH:
-                warnings.warn("Final registration failed, registration may be flawed")
+                warnings.warn("Registration failed, repeat with outlier removal")
+                t_matrix, min_dist = registration.particle_filter(
+                    fiducial_coords=fiducial_coords,
+                    spot_coords=spot_coords,
+                    particles=particles,
+                    stds=np.array(constants.STDS),
+                    stop_criteria=0.1,
+                    remove_outlier=True,
+                    debug=constants.DEBUG,
+                )
+                if min_dist > constants.REG_DIST_THRESH:
+                    warnings.warn("Final registration failed, registration may be flawed")
+                    well_failed.append(well_name)
```

Comment:
Why collect well names in a list if all you'll do is print them? You might as well print them as they happen. Ideally we should set all OD values to None for failed wells so no one mistakenly analyzes them.

Reply:
Throw warnings when the registration fails doesn't seem to work every time for some reasons... when I ran it in pycharm sometimes the registration clearly failed but no warning was output in the console. I'm not sure why it happened but printing the warning at the end seems to work more robustly. Also the pipeline prints lots of output (min distance from particle filtering in every interation) when run in debug mode, which makes it hard to track which well has failed. I think it would be useful to write this info in a log file or the output metadata, in addition to setting the spot intensity to none...

Reply:
I totally agree, I too have been thinking it would be great to add some logging. Hopefully it's something we can add once we have the more critical path things in place.